### ISSUE TITLE
Containers with unhealthy healthchecks don't reset to starting after container restart

### ISSUE THREAD

**Issue by shane-axiom** [docker/docker#27361]
**Description**

**Steps to reproduce the issue:**
1. Start a container with a HEALTHCHECK defined
2. Cause HEALTHCHECK to fail/container to become `unhealthy`
3. Restart the container

**Describe the results you received:**

Container is still `unhealthy` instead of `starting`.

**Describe the results you expected:**

Container should have `starting` status after restart.

If this is the intended behavior, is there another way to manually reset the container status back to `starting`? I'm dealing with some containers that take a while to initialize after a restart, during which they should be `starting` and not `unhealthy` from the previous run.

**Output of `docker version`:**

```
Docker version 1.12.1, build 23cf638
```

**Output of `docker info`:**

```
Containers: 6
 Running: 6
 Paused: 0
 Stopped: 0
Images: 6
Server Version: 1.12.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 88
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null bridge host overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options:
Kernel Version: 3.16.0-4-amd64
Operating System: Debian GNU/Linux 8 (jessie)
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 47.26 GiB
Name: SERVERNAME
ID: RIJZ:VQFC:B3AD:LRCQ:RCQ2:L7FG:RE7D:LD5S:6XAW:IZOG:UVFC:7DWZ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No kernel memory limit support
WARNING: No cpu cfs quota support
WARNING: No cpu cfs period support
```

**Comment by tonistiigi** [comment]
cc @talex5

**Comment by talex5** [comment]
Agree this looks like a bug. I've attempted a fix in #27387 if you have time to test it.

**Comment by shane-axiom** [comment]
@talex5 Thanks for fixing! Sorry, tried to build and test your branch but this is my first foray into building/testing docker from source. I was able to build the binaries with `make binary` but haven't had luck getting them to run. I can start the dockerd built by `make binary` but it produces a lot of errors like these:

```
ERRO[0006] libcontainerd: failed to receive event from containerd: rpc error: code = 13 desc = transport is closing
```

```
ERRO[0000] containerd: notify OOM events                 error=open /proc/26369/cgroup: no such file or directory
ERRO[0000] containerd: notify OOM events                 error=no init process found
```

and I'm unable to start any containers. I haven't found any docs about what I might be running into and many of the docs I found seem to be outdated (referencing `docker daemon -D` and no mention of containerd/runc). I don't want to hijack this issue with my debug troubles. If you have any quick tips or doc links I'll take a look, otherwise I tried!
